### PR TITLE
parse_float: Error when a float is attempted to be parsed into an invalid type

### PR DESCRIPTION
**Pull request by jvyden**
I experienced an error while writing some code:

```zig
var reset: u64 = try std.fmt.parseFloat(u64, resetStr);
```

Obviously, you can't parse a float into a u64, but the error message this gave confused me a little bit:

```zig
/home/jvyden/Documents/zig/build/stage4/lib/zig/std/fmt/parse_float/common.zig:89:17: error: reached unreachable code
        else => unreachable,
                ^~~~~~~~~~~
/home/jvyden/Documents/zig/build/stage4/lib/zig/std/fmt/parse_float/common.zig:57:31: note: called from here
        mantissa: mantissaType(T),
                  ~~~~~~~~~~~~^~~
```

I tried to fix this with by changing the unreachable to a `@compileError`, but [tiehuis had a better suggestion](https://github.com/ziglang/zig/pull/15593#issuecomment-1537265319). The error now prints a stack trace when calling `parseFloat` directly:

```zig
/home/jvyden/Documents/zig/build/stage4/lib/zig/std/fmt/parse_float/parse_float.zig:16:9: error: Cannot parse a float into a non-floating point type.
        @compileError("Cannot parse a float into a non-floating point type.");
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
referenced by:
    main: src/main.zig:4:31
    callMain: /home/jvyden/Documents/zig/build/stage4/lib/zig/std/start.zig:609:32
    remaining reference traces hidden; use '-freference-trace' to see all reference traces
```

### CODE REVIEW

**Comment by jvyden** [comment]
Do these test failures happen randomly or is it (somehow) something I did? I saw you retried the job, so I wasn't sure.

**Comment by jvyden** [comment]
OOPS

**Comment by Vexu** [comment]
I didn't do anything so it likely just failed randomly.

**Comment by jvyden** [comment]
Yeah, seems to be random - one of the ones that failed passed now. I cant seem to find any trace of something going wrong in the build log after looking for a little while.

**Comment by tiehuis** [comment]
I think it would make more sense to perform the type-check at the start of `parseFloat` as this is the only public function exposed by this library. Can keep all other branches unreachable.

https://github.com/ziglang/zig/blob/d71a43ec2c28a53a3e1d9bcb538707eca00a6fc0/lib/std/fmt/parse_float/parse_float.zig#L14

**Comment by jvyden** [comment]
Yeah, sounds like a better approach - I didn't think of publicity so I should have dug around some more, haha

I'll implement this in the next commit.

**Comment by jvyden** [comment]
Tried to neatly package everything into one commit but Git was fighting me. Apologies.

File: lib/std/fmt/parse_float/parse_float.zig [hunk 1]
```diff
@@ -12,6 +12,10 @@ pub const ParseFloatError = error{
 };
 
 pub fn parseFloat(comptime T: type, s: []const u8) ParseFloatError!T {
+    if (T != f16 and T != f32 and T != f64 and T != f128) {
```
Comment (squeek502) [inline lib/std/fmt/parse_float/parse_float.zig:15]:
```suggestion
    if (@typeInfo(T) != .Float) {
```

Comment (tiehuis) [reply lib/std/fmt/parse_float/parse_float.zig:15]: Don't yet support parsing `f80` which this would allow.

Comment (squeek502) [reply lib/std/fmt/parse_float/parse_float.zig:15]: Shouldn't that be a different compile error, though? `cannot parse a float into a non-floating point type` seems like a confusing compile error if you pass `f80` as the type. Maybe something like this instead?

```
if (@typeInfo(T) != .Float) {
    @compileError("cannot parse a float into a non-floating point type");
}
if (T == f80) {
    @compileError("TODO support parsing float to f80");
}
```

Comment (jvyden) [reply lib/std/fmt/parse_float/parse_float.zig:15]: Ah, didn't see `@typeInfo`. Learning a lot about Zig with this PR :wink:

Good catch on `f80` - do you usually open issues for TODOs relating to missing features such as this one or do they stay in code?

**Pull request merged**